File: Striver/DP/Q19.py
# 0/1 knapsack problem

# recursive solution
def knapsack(wt,val,idx,W):
    if idx == 0:
        if wt[idx] <= W:
            return val[idx]
        else:
            return 0
    notTake = knapsack(wt,val,idx-1,W)
    take = float('-inf')
    if wt[idx] <= W:
        take = val[idx] + knapsack(wt,val,idx-1,W-wt[idx])
    return max(notTake,take)

# memoization
def knapsackMemo(wt,val,idx,W,dp):
    if idx == 0:
        if wt[idx] <= W:
            return val[idx]
        else:
            return 0
    if dp[idx][W] != -1:
        return dp[idx][W]
    notTake = knapsackMemo(wt,val,idx-1,W,dp)
    take = float('-inf')
    if wt[idx] <= W:
        take = val[idx] + knapsackMemo(wt,val,idx-1,W-wt[idx],dp)
    dp[idx][W] = max(notTake,take)
    return dp[idx][W]

# tabulation
def knapsackTab(wt,val,W):
    n = len(wt)
    dp = [[0 for i in range(W+1)] for j in range(n)]
    for i in range(wt[0],W+1):
        dp[0][i] = val[0]
    
    for i in range(1,n):
        for j in range(1,W+1):
            notTake = dp[i-1][j]
            take = float('-inf')
            if wt[i] <= j:
                take = val[i] + dp[i-1][j-wt[i]]
            dp[i][j] = max(notTake,take)
    return dp[n-1][W]

# space optimization with 2 1D arrays
def knapsackSpaceOpt(wt,val,W):
    n = len(wt)
    dp = [0 for i in range(W+1)]
    curr = [0 for i in range(W+1)]
    for i in range(wt[0],W+1):
        dp[i] = val[0]
    
    for i in range(1,n):
        for j in range(1,W+1):
            notTake = dp[j]
            take = float('-inf')
            if wt[i] <= j:
                take = val[i] + dp[j-wt[i]]
            curr[j] = max(notTake,take)
        dp = curr
        curr = [0 for i in range(W+1)]
    return dp[W]

# space optimization with 1 1D array
def knapsackSpaceOpt2(wt,val,W):
    n = len(wt)
    dp = [0 for i in range(W+1)]
    for i in range(wt[0],W+1):
        dp[i] = val[0]
    
    for i in range(1,n):
        for j in range(W,wt[i]-1,-1):
            notTake = dp[j]
            take = float('-inf')
            if wt[i] <= j:
                take = val[i] + dp[j-wt[i]]
            dp[j] = max(notTake,take)
    return dp[W]

wt = [1,3,4,5]
val = [1,4,5,7]
W = 7
print(knapsack(wt,val,len(wt)-1,W))
dp = [[-1 for i in range(W+1)] for j in range(len(wt))]
print(knapsackMemo(wt,val,len(wt)-1,W,dp))
print(knapsackTab(wt,val,W))
print(knapsackSpaceOpt(wt,val,W))
print(knapsackSpaceOpt2(wt,val,W))

File: Striver/DP/Q19.java
package Striver.DP;

public class Q19{
    // 0/1 Knapsack Problem
    // Given weights and values of n items, put these items in a knapsack of capacity W to get the maximum total value in the knapsack.

    // recursive solution - Time: O(2^n) Space: O(n)
    public static int knapsack(int[] wt, int[] val, int idx, int W){
        if (idx == 0) {
            if (wt[idx] <= W) return val[idx];
            else return 0;
        }

        int notTake = knapsack(wt, val, idx-1, W);
        int take = Integer.MIN_VALUE;
        if (wt[idx] <= W) take = val[idx] + knapsack(wt, val, idx-1, W-wt[idx]);
        return Math.max(notTake, take);
    }

    // memoization - Time: O(n*W) Space: O(n*W)
    public static int knapsackMemo(int[] wt, int[] val, int idx, int W, int[][] dp){
        if (idx == 0) {
            if (wt[idx] <= W) return val[idx];
            else return 0;
        }

        if (dp[idx][W] != -1) return dp[idx][W];

        int notTake = knapsackMemo(wt, val, idx-1, W, dp);
        int take = Integer.MIN_VALUE;
        if (wt[idx] <= W) take = val[idx] + knapsackMemo(wt, val, idx-1, W-wt[idx], dp);
        return dp[idx][W] = Math.max(notTake, take);
    }

    // tabulation - Time: O(n*W) Space: O(n*W)
    public static int knapsackTab(int[] wt, int[] val, int idx, int W){
        int[][] dp = new int[idx+1][W+1];
        for (int i = wt[0]; i <= W; i++) dp[0][i] = val[0];
        for (int i = 1; i <= idx; i++) {
            for (int j = 0; j <= W; j++) {
                int notTake = dp[i-1][j];
                int take = Integer.MIN_VALUE;
                if (wt[i] <= j) take = val[i] + dp[i-1][j-wt[i]];
                dp[i][j] = Math.max(notTake, take);
            }
        }
        return dp[idx][W];
    }

    // space optimization with 2 rows - curr and dp - Time: O(n*W) Space: O(2*W)
    public static int knapsackTab2(int[] wt, int[] val, int idx, int W){
        int[] dp = new int[W+1];
        int[] curr = new int[W+1];
        for (int i = wt[0]; i <= W; i++) dp[i] = val[0];
        for (int i = 1; i <= idx; i++) {
            for (int j = 0; j <= W; j++) {
                int notTake = dp[j];
                int take = Integer.MIN_VALUE;
                if (wt[i] <= j) take = val[i] + dp[j-wt[i]];
                curr[j] = Math.max(notTake, take);
            }
            int[] temp = dp;
            dp = curr;
            curr = temp;
        }
        return dp[W];
    }

    // space optimization with 1 row - Time: O(n*W) Space: O(W)
    public static int knapsackTab3(int[] wt, int[] val, int idx, int W){
        int[] dp = new int[W+1];
        for (int i = wt[0]; i <= W; i++) dp[i] = val[0];
        for (int i = 1; i <= idx; i++) {
            for (int j = W; j >= 0; j--) {
                int notTake = dp[j];
                int take = Integer.MIN_VALUE;
                if (wt[i] <= j) take = val[i] + dp[j-wt[i]];
                dp[j] = Math.max(notTake, take);
            }
        }
        return dp[W];
    }

    public static void main(String[] args) {
        int[] wt = {1, 3, 4, 5};
        int[] val = {1, 4, 5, 7};
        int W = 7;
        int idx = wt.length-1;
        System.out.println(knapsack(wt, val, idx, W));
        int[][] dp = new int[idx+1][W+1];
        for (int[] row: dp) java.util.Arrays.fill(row, -1);
        System.out.println(knapsackMemo(wt, val, idx, W, dp));
        System.out.println(knapsackTab(wt, val, idx, W));
        System.out.println(knapsackTab2(wt, val, idx, W));
        System.out.println(knapsackTab3(wt, val, idx, W));
    }
}